narwhal/primary/src/synchronizer.rs

```diff
@@ -826,3 +863,161 @@ impl Synchronizer {
         Ok(self.inner.certificate_store.contains(&digest)?)
     }
 }
+
+/// Holds information for a suspended certificate. The certificate can be accepted into the DAG
+/// once `missing_parents` become empty.
+struct SuspendedCertificate {
+    certificate: Certificate,
+    missing_parents: HashSet<CertificateDigest>,
+    notify: broadcast::Sender<()>,
+}
+
+/// Keeps track of suspended certificates and their missing parents.
+/// The digest keys in `suspended` and `missing` can overlap, but a digest can exist in one map
+/// but not the other.
+///
+/// They can be combined into a single map, but it seems more complex to differentiate between
+/// suspended certificates that is not a missing parent of another, from a missing parent without
+/// the actual certificate.
+///
+/// Traversal of certificates that can be accepted should start from the missing map, i.e.
+/// 1. If a certificate exists in `missing`, remove its entry.
+/// 2. Find children of the certificate, update their missing parents.
+/// 3. If a child certificate no longer has missing parent, traverse from it with step 1.
```

Comment:
I would suggest adding an explicit comment that the implementation is not thread safe and to guarantee that mutations are consistent in this structure it will need to be guarded behind a mutex (as we currently do).

Reply:
This would be pretty obvious with compile errors from Rust?

Reply:
Added a comment on Synchronizer should access State via its methods.